Android/app/src/main/java/app/intra/net/doh/Race.java

```diff
@@ -33,38 +36,69 @@
   }
 
   private final List<Probe> probes;
-  private final Listener listener;
+  private Race(List<Probe> probes) {
+    this.probes = probes;
+  }
 
   /**
    * Creates a race between different servers.  To run the race, call start().
    * @param factory The factory will be called once for each server.
    * @param urls The URLs for all the DOH servers to compare.
    * @param listener Called once on an arbitrary thread with the result of the race.
    */
-  public Race(ServerConnectionFactory factory, String[] urls, Listener listener) {
-    probes = new ArrayList<>(urls.length);
+  static Race serverConnectionRace(ServerConnectionFactory factory,
+      String[] urls, Listener listener) {
+    List<Probe> probes = new ArrayList<>(urls.length);
+    for (int i = 0; i < urls.length; ++i) {
+      probes.add(new ServerConnectionProbe(factory, urls[i], new Callback(i, probes, listener)));
+    }
+    return new Race(probes);
+  }
+
+  /**
+   * Creates a race between different servers using Go.  To run the race, call start().
+   * @param context Used to read the IP addresses of the servers from storage.
+   * @param urls The URLs for all the DOH servers to compare.
+   * @param listener Called once on an arbitrary thread with the result of the race.
+   */
+  private static Race goRace(Context context, String[] urls, Listener listener) {
+    List<Probe> probes = new ArrayList<>(urls.length);
     for (int i = 0; i < urls.length; ++i) {
-      probes.add(new Probe(factory, urls[i], new Callback(i)));
+      probes.add(new GoProbe(context, urls[i], new Callback(i, probes, listener)));
     }
-    this.listener = listener;
+    return new Race(probes);
   }
 
-  public synchronized void start() {
-    for (Probe task : probes) {
-      task.start();
+  public static Race get(Context context, String[] urls, Listener listener) {
```

Comment:
Make the `start()` calls in this method too. No need to put it in a separate method. `Race` can be fully static.

Reply:
Done